vespajlib/src/main/java/com/yahoo/concurrent/classlock/ClassLocking.java

```diff
@@ -2,36 +2,40 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.function.Supplier;
 
 /**
  * @author valerijf
  */
 public class ClassLocking {
     private final Map<Class<?>, ClassLock> classLocks = new HashMap<>();
-    private final Object monitor = new Object();
 
-    public ClassLock lock(Class<?> clazz) {
-        synchronized (monitor) {
-            while(classLocks.containsKey(clazz)) {
-                try {
-                    monitor.wait();
-                } catch (InterruptedException ignored) { }
-            }
+    public synchronized ClassLock lock(Class<?> clazz) {
+        return tryLock(clazz, () -> true);
+    }
 
-            ClassLock classLock = new ClassLock(this, clazz);
-            classLocks.put(clazz, classLock);
-            return classLock;
+    public synchronized ClassLock tryLock(Class<?> clazz, Supplier<Boolean> continueRetrying) {
+        while(classLocks.containsKey(clazz)) {
+            try {
+                wait();
+            } catch (InterruptedException ignored) {
+                if (! continueRetrying.get()) {
```

Comment:
This need to be moved to after the catch. On notify()/notifyAll(), the wait() will just return - not throw an exception.

Reply:
Moved after `while`.

Reply:
It cannot be after the while.

Reply:
Moved after `catch`. :man_facepalming: